caffe2/operators/experimental/c10/cpu/add_cpu.cc

```diff
@@ -11,24 +11,27 @@ namespace {
 
 template <class DataType>
 void add_op_cpu_impl(
-    const Tensor& A,
-    const Tensor& B,
-    Tensor* C,
+    const C10Tensor& A_,
+    const C10Tensor& B_,
+    const C10Tensor& C_,
```

Comment:
You need to be careful here. These functions explicitly take an output tensor to fill in. That's OK, but you need to make sure what the output parameter is clear one way or another. In the Caffe2 style, this was done by passing in `Tensor*` versus `const Tensor&`. But you are not writing in the Caffe2 style anymore and now you've lost that information. It's really important to know what the outputs of an operator. I recommend you adopt the ATen style, which is:

1. Suffix your function name with `_out`, to make it clear that it writes an explicit output.
2. Put the output at the very *beginning* of the function signature, with the name `output`.

I don't know if this has implications for the Caffe2 bindings. If you want to propose an alternative format for explicit output we should talk about it.

Reply:
Yes, I agree we need to make this more clear. Pointers unfortunately don't really make sense anymore since we don't honor constness. Also, pointers make it harder to convert caffe2::Tensor to C10Tensor on the fly when calling the kernel (this is why I changed from pointer to reference):

    caffe2::Tensor output;
    add_cpu(..., C10Tensor(output));

The function name doesn't matter, it's ignored everywhere outside this file and the kernels are only called through the dispatcher, so an `_out` name doesn't do anything. Same for the names of parameters, they're only relevant inside this file.
I don't really have a good solution, but also don't want to block this diff on it.

Reply:
OK, so maybe the function name doesn't matter, but schema name matters, right? So if I write a schema for a function that explicitly takes output arguments, I want it to say `_out` in its name explicitly.

(By the way, I'm really not sure what the plan is for having actual schema structs is. I remember asking you about them, and you admitted they might not be a viable long term solution.)

(Also, there's is a well-known design problem in ATen/native, which is that there's a lot of boilerplate for writing both the nice functional and the `_out` accepting versions of functions. We don't know what to do about that. Absent a solution, we have to register functions for each of the variants.)

Reply:
Similarly, ordering of the output argument matters for schema as well.